src/core/router/transports/ssu/session.cc

```diff
@@ -361,89 +364,128 @@ void SSUSession::SendSessionRequest() {
+          return;  // TODO(anonimal): assert/throw?
+        }
+
+      // Compute exchanged session dataset size
+      bool const is_IPv6 = m_RemoteEndpoint.address().is_v6();
+
+      std::uint16_t const data_size =
+          DHKeySize::PubKey * 2  // DH X+Y
+          + message->GetIPAddressSize()  // Alice's address
+          + 2  // Alice's port
+          + (is_IPv6 ? 16 : 4)  // Bob's address
+          + 2  // Bob's port
+          + 4  // Alice's relay tag
+          + 4;  // Bob's signed-on time
+
+      // Create dataset of exchanged session data (the dataset Bob has signed)
+      core::OutputByteStream data(data_size);
+
+      // Our (Alice's) DH X
+      data.WriteData(m_DHKeysPair->public_key.data(), DHKeySize::PubKey);
+
+      // Bob's DH Y
+      data.WriteData(message->GetDhY(), DHKeySize::PubKey);
+
+      // Our (Alice's) IP and port
+      data.WriteData(message->GetIPAddress(), message->GetIPAddressSize());
+      data.Write<std::uint16_t>(message->GetPort());
+
+      // Bob's IP address
+      data.WriteData(
+          is_IPv6 ? m_RemoteEndpoint.address().to_v6().to_bytes().data()
+                  : m_RemoteEndpoint.address().to_v4().to_bytes().data(),
+          is_IPv6 ? 16 : 4);
+
+      // Bob's port
+      data.Write<std::uint16_t>(m_RemoteEndpoint.port());
+
+      // Our (Alice's) relay tag
+      data.Write<std::uint32_t>(m_RelayTag = message->GetRelayTag());
+
+      // Bob's signed-on time
+      data.Write<std::uint32_t>(message->GetSignedOnTime());
+
+      // Get Bob's padded signature length
+      std::uint8_t const signature_len =
+          SSUPacketBuilder::GetPaddedSize(m_RemoteIdentity.GetSignatureLen());
+
+      // Prepare decrypted-signature buffer
+      std::vector<std::uint8_t> signature(signature_len);
+
+      // Use Bob's IV to decrypt signature using our negotiated session key
+      m_SessionKeyDecryption.SetIV(message->GetHeader()->GetIV());
+
+      // Decrypt signature
+      m_SessionKeyDecryption.Decrypt(
+          message->GetSignature(), signature.size(), signature.data());
+
+      // TODO(anonimal): log debug of encrypted/decrypted sig + message data
+
+      // Verify signed dataset
+      if (!m_RemoteIdentity.Verify(data.Data(), data.Size(), signature.data()))
+        {
+          LOG(error) << "SSUSession:" << GetFormattedSessionInfo()
+                     << "SessionCreated signature verification failed";
+          // TODO(anonimal): review if Java routers resend the message on failure.
+          //   Instead of immediately resetting session key, we can explore ways
+          //   to observe and mitigate potential attacks. Another possible case
+          //   for failure:
+          //     "If Bob's NAT/firewall has mapped his internal port to a
+          //     different external port, and Bob is unaware of it, the
+          //     verification by Alice will fail."
+          m_IsSessionKey = false;
+          return;  // TODO(anonimal): throw/assert?
```

Comment:
Same question about throwing as for SendRelayResponse.
If retrying on failure, would an exception help signal to re-attempt the session?

Reply:
This will be addressed once a new message generator/dispatcher is implemented.